aiohttp/web_fileresponse.py

```diff
@@ -46,54 +53,54 @@ def _write(self, chunk: bytes) -> None:
         self.output_size += len(chunk)
         self._sendfile_buffer.append(chunk)
 
-    def _sendfile_cb(self, fut: 'asyncio.Future[None]',
-                     out_fd: int, in_fd: int,
-                     offset: int, count: int,
-                     loop: asyncio.AbstractEventLoop,
-                     registered: bool) -> None:
-        if registered:
-            loop.remove_writer(out_fd)
+    def _sendfile_cb(self, fut: 'asyncio.Future[None]', out_fd: int) -> None:
         if fut.cancelled():
             return
+        try:
+            if self._do_sendfile(out_fd):
+                set_result(fut, None)
+        except Exception as exc:
+            set_exception(fut, exc)
 
+    def _do_sendfile(self, out_fd: int) -> bool:
         try:
-            n = os.sendfile(out_fd, in_fd, offset, count)
+            n = os.sendfile(out_fd,
+                            self._in_fd,
+                            self._offset,
+                            self._count)
             if n == 0:  # EOF reached
-                n = count
+                n = self._count
         except (BlockingIOError, InterruptedError):
             n = 0
-        except Exception as exc:
-            set_exception(fut, exc)
-            return
+        self.output_size += n
+        self._offset += n
+        self._count -= n
+        return self._count <= 0
```

Comment:
Makes sense but I don't care too much

Reply:
I would prefer add this `assert` in order to be sure that count is always non-negative.

Reply:
@socketpair please don't use `assert`s in runtime! Only in tests. Use if+raise instead. The reason is that opcode for `assert` statement is ripped out in optimized mode (`-O`).

It works like this since forever:
* https://docs.python.org/2.0/ref/assert.html
* https://docs.python.org/3/reference/simple_stmts.html?highlight=assert#grammar-token-assert-stmt

Reply:
Well, yes. But it also brings misleading expectations, often affecting people's judgement resulting in relying on these checks more then allowed.

Reply:
Yes, I understand that assert may be optimized out. So, for that exact case, if `assert` proposed by me is optimized out it's OK. That condition should never been triggered actually. But if it is accidentally triggered in debug build (or, during tests) it is what I wanted.

If whole logic is correct, this check is superfluous and should not be checked in production.